### DISCUSSION TITLE
Help with setting up testing on my machine

### DISCUSSION THREAD

**Discussion by placoderm** [codewars/codewars.com#2608] [Q&A] [answered]
Please let me know if this is the right place for this question.

I'm a regular user, doing exclusively JavaScript challenges. I do all of my coding of the challenges in vscode on my device. I'd like to be able to run the tests in vscode as well. I am a little familiar with using Jest in React projects.

I can't figure out what testing package I should be installing, though. This is the challenge I'm working on now: https://www.codewars.com/kata/52742f58faf5485cae000b9a/train/javascript

This is what the testing block looks like:
```javascript
describe("Tests", () => {
  it("test", () => {
Test.assertEquals(formatDuration(1), "1 second");
  }); });
```
I have also seen this kind of test
```javascript
const chai = require("chai");
const assert = chai.assert;
chai.config.truncateThreshold=0;

describe("Basic tests", () => {
  it("Testing for fixed tests", () => {
    assert.strictEqual(solution('world'), 'dlrow');
  }); });  
```
I have read through [this document](https://docs.codewars.com/languages/javascript/) and I'm not quite seeing what testing package I should install. This challenge is Node v8.1.3. However the above document only talks about 10x and 12x. Should I just install Mocha and Chai? 

Does anyone have advice on the best VS Code plugins to use?

I'm very new at this so please assume I don't know what I'm doing 😬!

Thanks!

```

**Comment by kazk** [comment, marked as answer]
If you see `Test.` from old kata, you need [`@codewars/test-compat`](https://www.npmjs.com/package/@codewars/test-compat).

Install that with `npm i -D @codewars/test-compat` then:
```javascript
const Test = require("@codewars/test-compat");
describe("Tests", () => {
  it("test", () => {
    Test.assertEquals(formatDuration(1), "1 second");
  });
});
```
You should be able to run the test with Mocha.

Also, Codewars currently concatenates solution and tests, so write your solution in the same file.

**Reply by thesunshade** [reply]
Thanks!

When I run `npm i -D @codewars/javascript-test-compat` I get the following:
```
npm ERR! 404 Not Found - GET https://registry.npmjs.org/@codewars%2fjavascript-test-compat - Not foundnpm ERR! 404 
npm ERR! 404  '@codewars/javascript-test-compat@*' is not in this registry.
```
I have found this package on npm. is it the correct one?
https://www.npmjs.com/package/@codewars/test-compat
Should I instead be doing: `npm i -D @codewars/test-compat` ?

> Also, Codewars currently concatenates solution and tests, so write your solution in the same file.

Just to be clear, I'm only trying to solve the challenge, not create them.

**Reply by kazk** [reply]
Yeah, sorry, that's the correct package. I'll edit my answer.

**Reply by thesunshade** [reply]
No problem!

So in my working folder I have done 
`npm i -D @codewars/test-compat`
`npm install --global mocha`

I added this to the `packages.json` file:
```
"scripts": {
  "test": "mocha"
}
```

All that seemed to work fine. I have added the code you gave along with the test in the challenge I linked to in the OP. 

I then went into the terminal and ran `npm test`. This is the response I got:
![22-02-19 12_48_17-secondsToTimeDaysYears js - CodeWars - Visual Studio Code](https://user-images.githubusercontent.com/82448383/154791415-e08ec81f-c123-46dc-a08d-d2abe9fc6312.jpg)

So, then I tried moving the test into a file with the name `secondsToTimeDaysYears.test.js` as I think I would have done in Jest since the script I'm writing is in `secondsToTimeDaysYears.js`. Ran `npm test` and got the same response as above.

Then I tried just naming the file `test.js`. Ran `npm test` and this is the result:
![22-02-19 12_52_51-test js - CodeWars - Visual Studio Code](https://user-images.githubusercontent.com/82448383/154791421-a4df19b4-42e1-45ed-a0d4-2093012cdd49.jpg)


That's not so surprising since there is nothing connecting the test file with the challenge I'm doing. 

I also installed [Mocha Test Explorer](https://marketplace.visualstudio.com/items?itemName=hbenl.vscode-mocha-test-adapter) in VSCode, but that doesn't seem to be aware of the tests in any of the ways I tried above.

I realize that you can't provide tech support for Mocha, but if there is anything you or anyone else can suggest it will be most appreciated. BTW, I noticed that this post got marked as answered. I don't think I marked it as such, and it might get more input from others if it wasn't marked as answered. Not sure how it all works.

Thanks!!!

**Reply by kazk** [reply]
You didn't need to install `mocha` globally, `npm install -D mocha` should work because NPM scripts prepend `node_module/.bin/` to your `$PATH` (or whatever is equivalent on Windows).

To let Mocha find tests, put them in `test/`, name it `test.js`, or pass glob (`mocha *.test.js`).

To test, you need to write the solution in the same file as Codewars does, or add `module.exports = solutionName` to your solution file and import that in tests.

```javascript
// solution.js
function formatDuration(seconds) {
  //
}

module.exports = formatDuration;
```
```javascript
// test.js
const Test = require("@codewars/test-compat");
const formatDuration = require("./solution");
describe("Tests", () => {
  it("test", () => {
    Test.assertEquals(formatDuration(1), "1 second");
  });
});
```

**Reply by thesunshade** [reply]
> You didn't need to install `mocha` globally

Yes, I didn't think so, but the way the mocha docs were worded I thought it would be safer to do globally. Thanks so much for taking time to give such a detailed answer.

I'm still getting `ReferenceError: formatDuration is not defined`. 

I created a directory `test\` and moved the `test.js` file into it (same one in the screenshot above). I then put `module.exports = formatDuration;` in my solution file. 

![image](https://user-images.githubusercontent.com/82448383/154794095-cd6211b2-0a88-4fef-b802-52fc165c7618.png)


I'm not completely clear on how this:
> To test, you need to write the solution in the same file as Codewars does,

goes along with:
> To let Mocha find tests, put them in test/, name it test.js, 

I'm not even sure I understand what you mean by "write the solution in the same file as Codewars does". That my solution and the test need to be in the same file together??

I'm also trying to figure out how to "pass glob (`mocha *.test.js`)." It seems that this would be telling mocha to not only look for `test.js` for tests but anything that ends with `.test.js`, and that makes sense. Do I just enter that in the terminal one time, or does it need to go in a setting file somewhere?

I appreciate your patience and support. As you can see I'm quite new at all this, but I feel like I am already learning lots of things even though I don't have it working yet. Thanks!

EDIT: I also noticed in vscode that there were two gray dots under the `require`. Not sure if that means something:
![image](https://user-images.githubusercontent.com/82448383/154796826-e0c984a0-c208-4533-b9ee-fbba8111f4eb.png)

**Reply by kazk** [reply]
> I'm still getting `ReferenceError: formatDuration` is not defined.
> I'm not even sure I understand what you mean by "write the solution in the same file as Codewars does". That my solution and the test need to be in the same file together??

You need to import the solution like I wrote above (`const formatDuration = require("path/to/solution.js");`) to call that in tests. On Codewars, the test doesn't have that because it's prepended to the test file (same as writing in the same file).

That's why I wrote:

> To test, you need to write the solution in the same file as Codewars does, **OR** add `module.exports = solutionName` to your solution file and import that in tests.

---

> I'm also trying to figure out how to "pass glob (`mocha *.test.js`)." It seems that this would be telling mocha to not only look for `test.js` for tests but anything that ends with `.test.js`, and that makes sense. Do I just enter that in the terminal one time, or does it need to go in a setting file somewhere?

`mocha` accepts a list of test files. You can include the name of the test files in `package.json` if you want.

```
"scripts": {
  "test": "mocha your-test-file.js"
}
```

---

> EDIT: I also noticed in vscode that there were two gray dots under the require. Not sure if that means something

That's probably suggesting ES modules syntax (use `import` syntax). I don't have the time to explain Common JS (`require`) and ES Modules (`import`), but ES Modules is not the default at the moment, so it requires more steps to configure. I don't think it's a good idea for you.

**Reply by thesunshade** [reply]
> You need to import the solution like I wrote above (`const formatDuration = require("path/to/solution.js");`)

Oh darn! I missed that line. So sorry.
It's working now. Thank you so much for taking this time to help me. Hopefully this thread will be useful to other people too.

I really appreciate your patience.
